Implement StreamV2 to StreamV3 upgrading

mmathesius opened the pull request:
I think I have most of the low level functionality in place for upgrading StreamV2 to StreamV3.

Several of the tests generate interesting output, but don't really do much meaningful testing. I still need to work on those.

This low level functionality also needs to be integrated into the higher level stream upgrade functionality, but I'm still not exactly sure how to do that...

sgallagher reviewed (changes requested):
Not finished reviewing yet, but I wanted to give you some feedback.

mmathesius commented on modulemd/modulemd-module-stream.c:
```diff
@@ -651,6 +690,704 @@ modulemd_module_stream_upgrade_to_v2 (ModulemdModuleStream *from)
+        }
+
+
+      if (!g_strv_contains ((const gchar *const *)buildtime_modules,
+                            "platform") &&
+          !g_strv_contains ((const gchar *const *)runtime_modules, "platform"))
+        {
+          g_set_error_literal (error,
+                               MODULEMD_ERROR,
+                               MMD_ERROR_UPGRADE,
+                               "Stream v2 has no platform dependencies.");
+          return NULL;
+        }
```
> What should we do in cases of one or the other having a platform dep? For example, there could be modules built for F31 that have no runtime platform dep.

That case is handled by `stream_expansion_resolve_platform()`.

> I think we should treat "platform doesn't appear in deps" as equivalent to "dep is `[ ]` or negated".

Right. `[ ]` and negated streams get detected (and errored out) elsewhere. See the ""Cannot expand module %s dependency %s ..." errors in in `stream_expansion_helper()`. Perhaps a different error message is in order here?

mmathesius commented on modulemd/modulemd-module-stream.c:
```diff
@@ -651,6 +690,704 @@ modulemd_module_stream_upgrade_to_v2 (ModulemdModuleStream *from)
+              g_ptr_array_add (new_expanded_deps, g_steal_pointer (&new_dep));
+            }
+          /* otherwise, expand on what we already have */
+          else
+            {
+              /* for every expanded dependency we have so far... */
+              for (guint k = 0; k < (*expanded_deps)->len; k++)
+                {
+                  g_debug ("Expansion: expanding existing dependency");
+                  /* Make a copy of the existing expanded dependency and add this module and stream */
+                  new_dep = modulemd_build_config_copy (
+                    g_ptr_array_index (*expanded_deps, k));
+                  (*build_config_add_requirement) (new_dep, module, stream);
+                  g_ptr_array_add (new_expanded_deps,
+                                   g_steal_pointer (&new_dep));
+                }
+            }
```
This loop is the very heart of the stream expansion. It's iterative expanding the existing dependency set (`expanded_deps`) by making a copy of each item, adding the current module:stream as a dependency, then adding that updated copy to `new_expanded_deps`. `expanded_deps` stays unmodified until the end of the module loop when all the streams have been added as new dependences. Then it gets cleared and replaced by `new_expanded_deps`.  I'm sure I _could_ use `g_ptr_array_extend_and_steal()`, but I think it would mean creating yet another `GPtrArray` array, adding things to that, and then extending/stealing. I hate to make it even more difficult to understand!

sgallagher replied:
OK, I clearly didn't understand the algorithm well enough here. I think that's a good sign that it needs to have some heavy commenting done to explain things.

mmathesius replied:
Yeah. Stream expansion starts by generating a full Cartesian product of the StreamV2 dependencies to "flatten" them so they can be mapped to StreamV3. This is the guts of calculating that product and storing it in an array of Buildconfig objects. I'll work on adding commentary text.

mmathesius replied:
I just pushed a commit adding a lot of commentary. Hopefully it helps in understanding the code. (And it was good for me to try to explain it!)

mmathesius commented on modulemd/include/private/modulemd-module-stream-private.h:
```diff
@@ -382,4 +383,45 @@ gboolean
+ * Since: 2.10
+ */
+ModulemdModuleIndex *
+modulemd_module_stream_upgrade_v2_to_v3_ext (ModulemdModuleStream *from,
```
I just pushed a commit with that adjustment. (Along with your circular dependency fix. Thanks!)

mmathesius commented on modulemd/include/private/modulemd-module-stream-private.h:
```diff
@@ -382,4 +383,45 @@ gboolean
+ * have its dependencies expanded.
+ * @error: (out): A #GError that will return the reason for an expansion error.
+ *
+ * Stream V2 #ModulemdDependencies can have multiple streams specified per
+ * dependent buildtime and runtime modules. Stream V3 dependencies can only
+ * have a single stream per module. Additionally, Stream V2 dependencies treated
+ * "platform" the same as any other module dependency, while Stream V3
+ * dependencies have "platform" as a seperate property. This function expands
+ * the Stream V2 dependencies into the different possible combinations and
+ * explicitly sets the "platform".
```
`modulemd_module_stream_expand_v2_to_v3_deps()` passes `v2_stream` along to `stream_expansion_gen_contexts()`.

mmathesius commented on modulemd/include/modulemd-2.0/modulemd-module-stream.h:
```diff
@@ -275,6 +278,25 @@ modulemd_module_stream_upgrade (ModulemdModuleStream *self,
+ *
+ * Since: 2.10
+ */
+ModulemdModuleIndex *
```
@sgallagher It occurred to me that I can probably avoid some of the problems I was experiencing while trying to put my implementation of this function to use today by switching the return type to `ModulemdModule *`. It will avoid a lot of the overhead and internal complexity of using an index (such as it automatically attempting stream version upgrades at undesired times--which were causing me infinite recursion trouble). The automatic stream upgrades might even work. I'll give that a try tomorrow.

mmathesius commented:
I think I have most of the StreamV2 to StreamV3 upgrading integrated with my latest commit. I know there is still tricky work needed in `modulemd_module_add_obsoletes()`, but that may not be needed if we disallow automatic stream upgrades.

mmathesius commented:
FYI, my last force push is a rebase onto the current `mmdv3-devel` branch.

mmathesius commented:
My latest commit partially implements the API update that a ModuleIndex stream mdversion is no longer allowed to change on-the-fly. It's incomplete, but I got it back to a point that CI is passing again--so it is a good point to stop for the weekend.

mmathesius commented:
AFAIK, my latest commit pretty much wraps up the implementation.

The one remaining "TODO" I had flagged is what to do when `autogen_module_name` is `TRUE` when a reading a packager v3 document:
https://github.com/fedora-modularity/libmodulemd/blob/5ae727b2b7e343fc182060c48a361223ca86d36d/modulemd/modulemd-module-index.c#L194-L197

And, of course, many additional test cases should be added.

mmathesius commented:
My latest push rebases this PR to the latest `mmdv3-devel` branch, and adds a commit to adjust the tests that were added by PR #504 to avoid changing an the index stream mdversion on the fly.

mmathesius commented:
I added a few more commits to address the auto-generation of module/stream names and a few other things that popped up, and I also just rebased to the latest `mmdv3-devel` branch. Aside from adding more tests, this PR has everything I'm currently aware of.

sgallagher reviewed (approved):
I think we're good at this point. We can fix up any bugs that show up as they happen.

mmathesius commented:
My most recent force push made no code changes, but cleaned up the git commit history and comments.
